cmake/SpectreCheckTargetDependencies.cmake

```diff
@@ -0,0 +1,652 @@
+function(_absolute_source_paths TARGET_NAME)
+  get_property(
+    _SOURCE_FILES
+    TARGET ${TARGET_NAME}
+    PROPERTY SOURCES
+    )
+  get_property(
+    _LOCATION
+    TARGET ${TARGET_NAME}
+    PROPERTY FOLDER
+    )
+
+  unset(_ABS_SOURCE_FILES)
+  foreach(SOURCE ${_SOURCE_FILES})
+    if(IS_ABSOLUTE ${SOURCE})
+      list(APPEND _ABS_SOURCE_FILES ${SOURCE})
+    else()
+      list(APPEND _ABS_SOURCE_FILES "${_LOCATION}/${SOURCE}")
+    endif()
+  endforeach(SOURCE ${_SOURCE_FILES})
+  # "Return" SOURCE_FILES by setting it in PARENT_SCOPE
+  set(SOURCE_FILES ${_ABS_SOURCE_FILES} PARENT_SCOPE)
+endfunction(_absolute_source_paths TARGET_NAME)
+
+# Get the list of SpECTRE includes (i.e. not for 3rd party libraries)
+# for the target `TARGET_NAME`.
+#
+# Sets the variable `SPECTRE_INCLUDES` in parent scope to return the result
+function(_extract_spectre_includes FILE_NAME)
+  file(READ ${FILE_NAME} FILE_CONTENTS)
+  string(REGEX MATCHALL "\n#include \"[^\n]+\""
+    _RAW_SPECTRE_INCLUDES ${FILE_CONTENTS})
+  unset(SPECTRE_INCLUDES)
+  foreach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+    string(REPLACE "\n#include \"" "" INCLUDE "${INCLUDE}")
+    string(REPLACE "\"" "" INCLUDE "${INCLUDE}")
+    list(APPEND SPECTRE_INCLUDES ${INCLUDE})
+  endforeach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+  # "Return" SPECTRE_INCLUDE by setting it in PARENT_SCOPE
+  set(SPECTRE_INCLUDES ${SPECTRE_INCLUDES} PARENT_SCOPE)
+endfunction(_extract_spectre_includes FILE_NAME)
+
+# Get the list of 3rd party library includes for the target `TARGET_NAME`.
+#
+# Sets the variable `SPECTRE_INCLUDES` in parent scope to return the result
+function(_extract_tpl_includes FILE_NAME)
+  file(READ ${FILE_NAME} FILE_CONTENTS)
+  string(REGEX MATCHALL "\n#include <[^\n]+>"
+    _RAW_SPECTRE_INCLUDES ${FILE_CONTENTS})
+  unset(SPECTRE_INCLUDES)
+  foreach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+    string(REPLACE "\n#include <" "" INCLUDE "${INCLUDE}")
+    string(REPLACE ">" "" INCLUDE "${INCLUDE}")
+    list(APPEND SPECTRE_INCLUDES ${INCLUDE})
+  endforeach(INCLUDE ${_RAW_SPECTRE_INCLUDES})
+  # "Return" SPECTRE_INCLUDE by setting it in PARENT_SCOPE
+  set(SPECTRE_INCLUDES ${SPECTRE_INCLUDES} PARENT_SCOPE)
+endfunction(_extract_tpl_includes FILE_NAME)
+
+# Make the path of `HEADER_FILE` which belongs to the target `TARGET_NAME`
+# an absolute path.
+#
+# Sets the variable `ABS_HEADER_FILE` in parent scope to return the result
+function(_make_header_path_absolute TARGET_NAME HEADER_FILE)
+  get_property(
+    _INCLUDE_DIR
+    TARGET ${TARGET_NAME}
+    PROPERTY INTERFACE_INCLUDE_DIRECTORIES
+    )
+  if(IS_ABSOLUTE ${HEADER_FILE})
+    set(_ABS_HEADER_FILE ${HEADER_FILE})
+  else()
+    set(_ABS_HEADER_FILE "${_INCLUDE_DIR}/${HEADER_FILE}")
+  endif()
+  # "Return" ABS_HEADER_FILE by setting it in PARENT_SCOPE
+  set(ABS_HEADER_FILE ${_ABS_HEADER_FILE} PARENT_SCOPE)
+endfunction(_make_header_path_absolute TARGET_NAME HEADER_FILE)
+
+# Gets the dependencies for a list of includes.
```

Comment:
Which list of includes? can it be arbitrary? should it be the includes from one file? it's unclear why the function is set up this way (in C++ I would have a function to get the deps for 1 include, and another function wrapping it to get the deps from a file, but of course good C++ isn't good cmake). It would be helpful to document how this function is intended to be used.

Reply:
This function gets all dependencies for all includes from one file. You can't really `return` things in CMake and so control flow is horrid, which is why it is written this way.